### ISSUE TITLE
How/when are models registered?

### ISSUE THREAD

**Issue by dodas** [xaviergonz/mobx-keystone#183]
Hi! First of all, thank you for mobx-keystone. The TS issues in MST are quite bothering.

I run into an error when trying to deserialize snapshot:
```
Error: model with name "app/User" not found in the registry
```

This is the gist of the code:
```ts
// store/User.ts

@model("app/User")
export class User extends Model({
  name: prop<string>(),
}) {
}

// store/RootStore.ts
import { User } from "./User";

@model("app/RootStore")
export class RootStore extends Model({
  user: prop<User>(),
}) {
}

// store/index.ts:

import { User } from "./User";
import { RootStore } from "./RootStore";

export function initStore(snapshot: SnapshotInOf<RootStore>) {
  // uncommenting following line solves the error
  // new User({});
  const store = fromSnapshot<RootStore>(snapshot);
  registerRootStore(store);
  return store;
}
```

Is this expected behaviour? Any ideas about how to overcome this in a better way?
Thanks!

**Comment by xaviergonz** [comment]
That's because ts is being smart and removing the User import when nothing from it gets used

https://devblogs.microsoft.com/typescript/announcing-typescript-3-8/#type-only-imports-exports

Try``` import "./User"``` in store index instead

**Comment by xaviergonz** [comment]
And maybe the same for RootStore actually. You shouldn't need this once you actually start using the models in your code though

**Comment by dodas** [comment]
@xaviergonz thanks for the feedback.

Could I workaround this by using runtime checking, therefore referencing `User` class as value, not only type?

Something like:
```ts
import { User } from "./User";

@model("app/RootStore")
export class RootStore extends Model({
  // Why does following line throw this TS error?

  // Type 'User' does not satisfy the constraint 'AnyType'.
  // Property '$$recordType' is missing in type 'User' but required in type 'RecordType<any>'

  // Referencing User class as value here, so hopefully it doesn't get tree-shaken.
  user: tProp<User>(types.model(User)),
} {}

**Comment by xaviergonz** [comment]
Yes, that would work. But just using ```new User``` anywhere in the code base should achieve the same

**Comment by dodas** [comment]
Could you explain why I am getting that TS error when trying to use runtime validation for class model?
```
Type 'User' does not satisfy the constraint 'AnyType'.
Property '$$recordType' is missing in type 'User' but required in type 'RecordType<any>'
```

**Comment by xaviergonz** [comment]
Could you paste where that error happens?

**Comment by TobiasWehrum** [comment]
> You shouldn't need this once you actually start using the models in your code though

My client app only uses models via `fromSnapshot`, so there are legitimate use cases where the `Error: model with name "app/User" not found in the registry` problem never just fixes itself.

This thread helped me (I'm using `import "./User"` for all models now, which feels a bit hacky, but at least it works), but I think this would be a good idea to mention in the docs.

As a side note, `"importsNotUsedAsValues" : "preserve"` in the tsconfig.json should also help, but something about my TypeScript-babel-webpack setup (I'm using https://github.com/gilamran/fullstack-typescript as a base) prevents it from working on the client side of the app.

**Comment by TrevorPage** [comment]
Having just encountered the same issue as the OP, the slightly hacky way I'm getting around it is:

```
import { OSCModulator } from './OSCModulator';

...

OSCModulator.name;
```

**Comment by TrevorPage** [comment]
Thank you @TobiasWehrum ! The only issue I have is with Visual Studio Code automatically removing unused imports, which I find quite useful. Therefore I added the extra reference(s) to circumvent that. Or, do you still think I might be doing something unnecessary there?
Thank you again for responding! Glad to meet others using this amazing library. It's incredible what I've managed to achieve with it in just a single evening.

**Comment by TobiasWehrum** [comment]
@TrevorPage I mean, if you want to keep it from removing the import, that's fair - the only downside that I can see is that it's two lines per model instead of just one, so no problem at all.

(Also, agreed. This is a fantastic library!)

**Comment by xaviergonz** [comment]
I put together a fix in PR #571: https://github.com/xaviergonz/mobx-keystone/pull/571

What this adds:
- A new `registerModels(...)` helper you can call at app startup to make model registration explicit when needed.
- Better error messages when a `$modelType` can’t be found in the registry, with concrete remediation steps.
- Docs updates explaining when this is usually *not* needed and when it *is* needed.

Typical usage:

```ts
import { registerModels } from "mobx-keystone"
import { RootStore } from "./RootStore"
import { User } from "./User"

registerModels(RootStore, User)
```

This is mostly useful for snapshot-heavy apps where imports can be elided (for example type-only imports). In many apps you won’t need it, because normal runtime imports already evaluate the model modules and register them.

Alternatively, it is now also possible to pass the model to fromSnapshot.
```
fromSnapshot(User, userSn)
```
Using it like that will just trigger registration too.
